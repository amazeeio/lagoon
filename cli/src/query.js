// @flow

import path from 'path';
import os from 'os';
import url from 'url';
import R from 'ramda';
import { config } from './config/';
import { fileExists, readFile } from './util/fs';
import request from './util/request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = path.join(os.homedir(), '.lagoon-token');
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const apiUrl = R.prop('api', config) || 'https://api.amazee.io/graphql';

  const {
    hostname, path: pathname, port: urlPort, protocol,
  } = url.parse(
    apiUrl,
  );

  if (hostname == null) {
    throw new Error(
      'API URL configured under the "api" key in .lagoon.yml doesn\'t contain a valid hostname.',
    );
  }

  const body = JSON.stringify(
    {
      query,
      variables,
    },
    null,
    pretty ? 2 : 0,
  );

  const protocolPorts = {
    'https:': 443,
    'http:': 80,
  };

  const options = {
    hostname,
    path: pathname,
    port:
      urlPort === null
        ? R.propOr(443, protocol)(protocolPorts)
        : Number(urlPort),
    method: 'POST',
    headers,
    body,
    rejectUnauthorized: false,
  };

  try {
    return await request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always('Could not connect to API.'),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }
}

export default runGQLQuery;
