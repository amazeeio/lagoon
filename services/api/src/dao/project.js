const R = require('ramda');
const {
  ifNotAdmin,
  inClauseOr,
  knex,
  prepare,
  query,
  whereAnd,
  isPatchEmpty,
} = require('./utils');

// This contains the sql query generation logic
const Sql = {
  updateProject: (cred, input) => {
    const { id, patch } = input;
    const { projects } = cred.permissions;

    const ret = knex('project')
      .where('id', '=', id)
      .whereIn('id', projects)
      .update(patch);

    return ret.toString();
  },
  selectProject: id =>
    knex('project')
      .where('id', id)
      .toString(),
  selectProjectIdByName: name =>
    knex('project')
      .where('name', name)
      .select('id')
      .toString(),
};

const Helpers = {
  getProjectIdByName: async (sqlClient, name) => {
    const pidResult = await query(sqlClient, Sql.selectProjectIdByName(name));

    const amount = R.length(pidResult);
    if (amount > 1) {
      throw new Error(
        `Multiple project candidates for '${name}' (${amount} found). Do nothing.`,
      );
    }

    if (amount === 0) {
      throw new Error(`Not found: '${name}'`);
    }

    const pid = R.path(['0', 'id'], pidResult);

    return pid;
  },
};

const getAllProjects = ({ sqlClient }) => async (cred, args) => {
  const { customers, projects } = cred.permissions;

  // We need one "WHERE" keyword, but we have multiple optional conditions
  const where = whereAnd([
    args.createdAfter ? 'created >= :createdAfter' : '',
    args.gitUrl ? 'git_url = :gitUrl' : '',
    ifNotAdmin(
      cred.role,
      `(${inClauseOr([['customer', customers], ['project.id', projects]])})`,
    ),
  ]);

  const prep = prepare(sqlClient, `SELECT * FROM project ${where}`);
  const rows = await query(sqlClient, prep(args));

  return rows;
};

const getProjectByEnvironmentId = ({ sqlClient }) => async (cred, eid) => {
  if (cred.role !== 'admin') {
    throw new Error('Unauthorized');
  }
  const prep = prepare(
    sqlClient,
    `SELECT
        p.*
      FROM environment e
      JOIN project p ON e.project = p.id
      WHERE e.id = :eid
    `,
  );

  const rows = await query(sqlClient, prep({ eid }));

  return rows ? rows[0] : null;
};

const getProjectByGitUrl = ({ sqlClient }) => async (cred, args) => {
  const { customers, projects } = cred.permissions;
  const str = `
      SELECT
        *
      FROM project
      WHERE git_url = :gitUrl
      ${ifNotAdmin(
    cred.role,
    `AND (${inClauseOr([
      ['customer', customers],
      ['project.id', projects],
    ])})`,
  )}
      LIMIT 1
    `;

  const prep = prepare(sqlClient, str);
  const rows = await query(sqlClient, prep(args));

  return rows ? rows[0] : null;
};

const getProjectByName = ({ sqlClient }) => async (cred, args) => {
  const { customers, projects } = cred.permissions;
  const str = `
      SELECT
        *
      FROM project
      WHERE name = :name
      ${ifNotAdmin(
    cred.role,
    `AND (${inClauseOr([
      ['customer', customers],
      ['project.id', projects],
    ])})`,
  )}
    `;

  const prep = prepare(sqlClient, str);

  const rows = await query(sqlClient, prep(args));

  return rows[0];
};

const addProject = ({ sqlClient }) => async (cred, input) => {
  const { customers } = cred.permissions;
  const cid = input.customer.toString();

  if (cred.role !== 'admin' && !R.contains(cid, customers)) {
    throw new Error('Project creation unauthorized.');
  }

  const prep = prepare(
    sqlClient,
    `CALL CreateProject(
        :id,
        :name,
        :customer,
        :git_url,
        ${input.subfolder ? ':subfolder' : 'NULL'},
        :openshift,
        ${
  input.openshift_project_pattern
    ? ':openshift_project_pattern'
    : 'NULL'
},
        ${
  input.active_systems_deploy
    ? ':active_systems_deploy'
    : '"lagoon_openshiftBuildDeploy"'
},
        ${
  input.active_systems_promote
    ? ':active_systems_promote'
    : '"lagoon_openshiftBuildDeploy"'
},
        ${
  input.active_systems_remove
    ? ':active_systems_remove'
    : '"lagoon_openshiftRemove"'
},
        ${input.branches ? ':branches' : '"true"'},
        ${input.pullrequests ? ':pullrequests' : '"true"'},
        ${input.production_environment ? ':production_environment' : 'NULL'},
        ${input.auto_idle ? ':auto_idle' : '1'},
        ${input.storage_calc ? ':storage_calc' : '1'}
      );
    `,
  );

  const rows = await query(sqlClient, prep(input));
  const project = R.path([0, 0], rows);

  return project;
};

const deleteProject = ({ sqlClient }) => async (cred, input) => {
  const { projects } = cred.permissions;

  // Will throw on invalid conditions
  const pid = await Helpers.getProjectIdByName(sqlClient, input.project);

  if (cred.role !== 'admin') {
    if (!R.contains(pid, projects)) {
      throw new Error('Unauthorized.');
    }
  }

  const prep = prepare(sqlClient, 'CALL DeleteProject(:project)');
  await query(sqlClient, prep(input));

  // TODO: maybe check rows for changed result
  return 'success';
};

const updateProject = ({ sqlClient }) => async (cred, input) => {
  const { projects } = cred.permissions;
  const pid = input.id.toString();

  if (cred.role !== 'admin' && !R.contains(pid, projects)) {
    throw new Error('Unauthorized');
  }

  if (isPatchEmpty(input)) {
    throw new Error('input.patch requires at least 1 attribute');
  }

  await query(sqlClient, Sql.updateProject(cred, input));
  const rows = await query(sqlClient, Sql.selectProject(pid));
  const project = R.path([0], rows);

  return project;
};

const Queries = {
  deleteProject,
  addProject,
  getProjectByName,
  getProjectByGitUrl,
  getProjectByEnvironmentId,
  getAllProjects,
  updateProject,
};

module.exports = {
  Sql,
  Queries,
  Helpers,
};
