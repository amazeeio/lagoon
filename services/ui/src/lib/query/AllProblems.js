import gql from 'graphql-tag';

export default gql`
  {
    allGroups {
      name
      projects {
        name
        id
        environments(type: PRODUCTION) {
          name
          id
          problems {
            id
            identifier
            severity
            source
          }
        }
      }
    }
  }
`;
