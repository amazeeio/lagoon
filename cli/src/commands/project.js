// @flow

import { table } from 'table';
import R from 'ramda';

import { visit } from '../cli';
import { config } from '../config';
import gql from '../gql';
import { runGQLQuery } from '../query';
import {
  printGraphQLErrors,
  printProjectConfigurationError,
} from '../printErrors';

import typeof Yargs from 'yargs';
import type { BaseArgs } from '.';

export const command = 'project';
export const description = 'Show project details';

export function builder(yargs: Yargs): Yargs {
  return yargs
    .usage(`$0 ${command} - ${description}`)
    .options({
      project: {
        demandOption: false,
        describe: 'Name of project',
        type: 'string',
      },
    })
    .alias('p', 'project')
    .example(
      `$0 ${command}`,
      'Show details for the project configured in .lagoon.yml',
    )
    .example(`$0 ${command} myproject`, 'Show details of project "myproject"')
    .commandDir('projectCommands', { visit });
}

type projectDetailsArgs = {
  projectName: string,
  clog: typeof console.log,
  cerr: typeof console.error,
};

export async function projectDetails({
  projectName,
  clog,
  cerr,
}:
projectDetailsArgs): Promise<number> {
  const query = gql`
    query ProjectByName($project: String!) {
      projectByName(name: $project) {
        name
        customer {
          name
        }
        git_url
        active_systems_deploy
        active_systems_remove
        branches
        pullrequests
        openshift {
          name
        }
        created
      }
    }
  `;

  const result = await runGQLQuery({
    cerr,
    query,
    variables: { project: projectName },
  });

  const { errors } = result;
  if (errors != null) {
    return printGraphQLErrors(cerr, ...errors);
  }

  const project = R.path(['data', 'projectByName'])(result);

  if (project == null) {
    clog(`No project '${projectName}' found.`);
    return 0;
  }

  clog(`Project details for '${projectName}':`);
  clog(
    table([
      ['Project Name', R.prop('name', project)],
      ['Customer', R.path(['customer', 'name'], project)],
      ['Git URL', R.prop('git_url', project)],
      ['Active Systems Deploy', R.prop('active_systems_deploy', project)],
      ['Active Systems Remove', R.prop('active_systems_remove', project)],
      ['Branches', String(R.prop('branches', project))],
      ['Pull Requests', String(R.prop('pullrequests', project))],
      ['Openshift', R.path(['openshift', 'name'], project)],
      ['Created', R.path(['created'], project)],
    ]),
  );

  return 0;
}

type Args = BaseArgs & {
  argv: {
    project: ?string,
  },
};

export async function handler({ clog, cerr, argv }: Args): Promise<number> {
  const projectName = R.prop('project', argv) || R.prop('project', config);

  if (projectName == null) {
    return printProjectConfigurationError(cerr);
  }

  return projectDetails({ projectName, clog, cerr });
}
