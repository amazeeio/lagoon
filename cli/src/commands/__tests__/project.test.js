// @flow

import { runGQLQuery } from '../../query';
import { projectDetails } from '../project';

jest.mock('../../query');

const _mock = (mockFn: any): JestMockFn => mockFn;

const mockErrorResponse = {
  errors: [{ message: 'something something error' }],
};

describe('projectDetails', () => {
  const mockResponse1 = {
    data: {
      projectByName: {
        name: 'credentialstest-project1',
        customer: {
          name: 'credentialtest-customer1',
        },
        git_url: 'project1.git',
        active_systems_deploy: 'lagoon_openshiftBuildDeploy',
        active_systems_remove: 'lagoon_openshiftRemove',
        branches: 'true',
        pullrequests: null,
        openshift: {
          name: 'credentialtest-openshift',
        },
        created: '2018-01-15 11:09:35',
      },
    },
  };

  it('should display error, if GraphQL sends error messages', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockErrorResponse),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await projectDetails({
      projectName: 'some_project',
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should show error on missing project', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve({}));

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await projectDetails({
      projectName: 'not_existing',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should show details for given project', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockResponse1),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await projectDetails({
      projectName: 'myproject',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
