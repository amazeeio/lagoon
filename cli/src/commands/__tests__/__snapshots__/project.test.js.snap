// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectDetails should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The lagoon API returned an error:[39m",
  ],
  Array [
    "[31m{[39m
[31m  \\"message\\": \\"something something error\\"[39m
[31m}[39m",
  ],
]
`;

exports[`projectDetails should show details for given project 1`] = `
Array [
  Array [
    "Project details for 'myproject':",
  ],
  Array [
    "╔═══════════════════════╤═════════════════════════════╗
║ Project Name          │ credentialstest-project1    ║
╟───────────────────────┼─────────────────────────────╢
║ Customer              │ credentialtest-customer1    ║
╟───────────────────────┼─────────────────────────────╢
║ Git URL               │ project1.git                ║
╟───────────────────────┼─────────────────────────────╢
║ Active Systems Deploy │ lagoon_openshiftBuildDeploy ║
╟───────────────────────┼─────────────────────────────╢
║ Active Systems Remove │ lagoon_openshiftRemove      ║
╟───────────────────────┼─────────────────────────────╢
║ Branches              │ true                        ║
╟───────────────────────┼─────────────────────────────╢
║ Pull Requests         │ null                        ║
╟───────────────────────┼─────────────────────────────╢
║ Openshift             │ credentialtest-openshift    ║
╟───────────────────────┼─────────────────────────────╢
║ Created               │ 2018-01-15 11:09:35         ║
╚═══════════════════════╧═════════════════════════════╝
",
  ],
]
`;

exports[`projectDetails should show error on missing project 1`] = `
Array [
  Array [
    "No project 'not_existing' found.",
  ],
]
`;
